src/main/java/com/jnape/palatable/lambda/semigroup/builtin/Absent.java

```diff
@@ -51,4 +55,37 @@ private Absent() {
     public static <A> Maybe<A> absent(Semigroup<A> semigroup, Maybe<A> x, Maybe<A> y) {
         return absent(semigroup, x).apply(y);
     }
+
+    private static <A> Semigroup<Maybe<A>> shortCircuitSemigroup(Semigroup<A> aSemigroup) {
+        return new Semigroup<Maybe<A>>() {
+            @Override
+            public Maybe<A> checkedApply(Maybe<A> maybeX, Maybe<A> maybeY) {
+                return LiftA2.liftA2(aSemigroup, maybeX, maybeY);
+            }
+
+            @Override
+            public Maybe<A> foldLeft(Maybe<A> aMaybe, Iterable<Maybe<A>> maybes) {
+                Maybe<A> accumulation = aMaybe;
+                for (Maybe<A> a : maybes) {
+                    boolean shouldShortCircuit = accumulation == Maybe.nothing();
+                    if (shouldShortCircuit)
+                        return accumulation;
+                    accumulation = checkedApply(accumulation, a);
+                }
+                return accumulation;
```

Comment:
Let's use a trampoline here to avoid the mutable `accumulation` ref:

```java
return trampoline(
        into((res, it) -> res.equals(nothing()) || !it.hasNext()
                          ? terminate(res)
                          : recurse(tuple(liftA2(aSemigroup, res, it.next()), it))),
        tuple(acc, maybes.iterator()));
```

Reply:
done